compiler/src/iree/compiler/Dialect/Stream/Transforms/ElideTransferOps.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/Stream/Analysis/Affinity.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-stream-elide-transfer-ops"
+
+namespace mlir::iree_compiler::IREE::Stream {
+
+#define GEN_PASS_DEF_ELIDETRANSFEROPSPASS
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h.inc"
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Transfer Elision Patterns
+//===----------------------------------------------------------------------===//
+
+struct ElideCompatibleTransferPattern
+    : public OpRewritePattern<IREE::Stream::AsyncTransferOp> {
+  AffinityAnalysis &affinityAnalysis;
+
+  ElideCompatibleTransferPattern(MLIRContext *context,
+                                 AffinityAnalysis &affinityAnalysis)
+      : OpRewritePattern(context), affinityAnalysis(affinityAnalysis) {}
+
+  LogicalResult matchAndRewrite(IREE::Stream::AsyncTransferOp transferOp,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(
+        { llvm::dbgs() << "[elide-transfers] Op: " << transferOp << "\n"; });
+
+    auto source = transferOp.getSource();
+    auto result = transferOp.getResult();
+    auto sourceType = cast<IREE::Stream::ResourceType>(source.getType());
+    auto resultType = cast<IREE::Stream::ResourceType>(result.getType());
+
+    // Don't elide transfers that change lifetime (usage casts)
+    if (sourceType.getLifetime() != resultType.getLifetime()) {
+      LLVM_DEBUG({
+        llvm::dbgs()
+            << "[elide-transfers] not eliding lifetime-changing transfer: "
+            << sourceType.getLifetime() << " -> " << resultType.getLifetime()
+            << "\n";
+      });
+      return failure();
+    }
+
+    // Get source and result affinities, either from explicit attributes or
+    // from affinity analysis.
+    auto sourceAffinityAttr = transferOp.getSourceAffinityAttr();
+    auto resultAffinityAttr = transferOp.getResultAffinityAttr();
+
+    if (!sourceAffinityAttr) {
+      sourceAffinityAttr = affinityAnalysis.lookupResourceAffinity(source);
+    }
+    if (!resultAffinityAttr) {
+      resultAffinityAttr = affinityAnalysis.lookupResourceAffinity(result);
+    }
+
+    if (!sourceAffinityAttr || !resultAffinityAttr) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "[elide-transfers] missing affinity information\n";
+      });
+      return failure();
```

Comment:
If a pattern fails to match return with `notifyMatchFailure` instead of printing on your own - this ensures that the `pattern-application` debug prints get meaningful messages in them and threads them correctly. In general, you should never `return failure();` from a pattern, and you should avoid debug prints unless there's no other way to do it. Generally, if your patterns are complex enough that "didn't match for reason X" is insufficient then it probably shouldn't be a pattern and instead be a proper pass.

```suggestion
      return rewriter.notifyMatchFailure(op, "missing affinity information");
```